src/kernel-install/ukify.py

```diff
@@ -0,0 +1,233 @@
+        offset = int(parts[2], base=0) if len(parts) >= 3 else None
+        return cls.create(name, content, offset)
+
+
+def parse_args():
+    p = argparse.ArgumentParser(
+        description='Build and sign Unified Kernel Images',
+        usage='''\
+usage: ukify [options…] linux initrd
+       ukify -h | --help
+''')
+
+    # Suppress printing of usage synopsis on errors
+    p.error = lambda message: p.exit(2, f'{p.prog}: error: {message}\n')
+
+    p.add_argument('linux',
+                   type=pathlib.Path)
+    p.add_argument('initrd',
+                   type=pathlib.Path)
+
+    p.add_argument('--cmdline')
+
+    p.add_argument('--os-release',
+                   type=pathlib.Path)
+
+    p.add_argument('--devicetree',
+                   type=pathlib.Path)
+    p.add_argument('--splash',
+                   type=pathlib.Path)
+
+    p.add_argument('--efi-arch',
+                   choices=('ia32', 'x64', 'arm', 'aa64', 'riscv64'))
+
+    p.add_argument('--stub',
+                   type=pathlib.Path)
+
+    p.add_argument('--section',
+                   dest='sections',
+                   metavar='NAME:CONTENTS[:OFFSET]',
+                   type=Section.parse_arg,
+                   action='append')
+
+    p.add_argument('--output', '-o',
+                   nargs='?')
+
+    p.add_argument('--key',
+                   type=pathlib.Path)
+    p.add_argument('--cert',
+                   type=pathlib.Path)
+
+    p.add_argument('--measure',
+                   action='store_true')
+
+    args = p.parse_args()
+
+    if bool(args.key) ^ bool(args.cert):
+        raise ValueError('--key and --cert must be specified together')
+
+    if args.os_release is None:
+        p = pathlib.Path('/etc/os-release')
+        if not p.exists():
+            p = pathlib.Path('/usr/lib/os-release')
+        args.os_release = p
+
+    if args.efi_arch is None:
+        args.efi_arch = guess_efi_arch()
+
+    if args.stub is None:
+        args.stub = f'/usr/lib/systemd/boot/efi/linux{args.efi_arch}.efi.stub'
+
+    if args.output is None:
+        suffix = '.efi' if args.key else '.unsigned.efi'
+        args.output = args.linux.name + suffix
+
+    return args
+
+
+def check_inputs(args):
+    for name, value in vars(args).items():
+        if name == 'output':
+            continue
+
+        if not isinstance(value, pathlib.Path):
+            continue
+
+        # Open file to check that we can read it, or generate an exception
+        open(value).close()
+
+
+def make_uki(args):
+    sections = [
+        # FIXME: calculate offsets based on content size
+        ('osrel',   args.os_release,    0x20_000),
+        ('cmdline', args.cmdline,       0x30_000),
+        ('dtb',     args.devicetree,    0x40_000),
+        ('splash',  args.splash,       0x100_000),
+        ('initrd',  args.initrd,     0x3_000_000),
+        # linux shall be last to leave breathing room for decompression
+        ('linux',   args.linux,      0x2_000_000),
```

Comment:
Reordering this here without giving it a different offset is not gonna do anything. And you cannot put it last without knowing the initrd size anyway.

Also, now that I think of this: it probably makes sense to put the linux section first and the osrel second section offset a suitable gap. That would guarantee there is a gap that does not contain anything important in it (all other sections are technically optional). If the linux section is last, the decompressor could conceivably override something important…

Reply:
> Why do we actually the sections like splash and dtb (which we wouldn't decompress)?

I can't parse this.

Basically, you look at the output of `objdump -h` or use the `pefile` python module to get at the data without ugly text parsing and then compute the offsets like this:

```
base_offset=highest_vma + highest_vma_size

if arch == x86 and kernel_PE_MajorImageVersion < 1:
    kernel_section_offset=align(base_offset, kernel_PE_OptionalHeader_SectionAlignment)
    kernel_size = MAX(kernel_file_size, kernel_PE_OptionalHeader_SizeOfCode)
else:
    kernel_section_offset=base_offset
    kernel_size = kernel_file_size

section1_offset=kernel_section_offset + kernel_size
section2_offset=section1_offset + section1_size
section3_offset=section2_offset + section2_size
…
```